### ISSUE TITLE
Listen for server ping event

### ISSUE THREAD

**Issue by mhstoller** [dart-lang/web_socket_channel#124]
Need a way to listen for ping events sent from a server. Something like
```
        channel.stream.listen(
        (message) {
          print("recieved: $message");
        },
        onPing: () {
          //reset ping timer
        },
        onError: (err) {
          print(err);
        },
        onDone: () {
          print("ws closed");
        });
```

The library includes a pingInterval property for the websocket to be able to ping the server, but my server is already pinging the client on a fixed interval, so sending pings from the client as well is redundant. What I want is the ability to tell from the client when a server ping has been received.

**Comment by ffactory-ofcl** [comment]
Hey I might be late but reading this I thought: It's not redundant to send pings from client to server as well as the other direction. If you don't, how do you know when the client disconnected? You're not sending pings you must wait for the underlying connection times out.

**Comment by mhstoller** [comment]
> Hey I might be late but reading this I thought: It's not redundant to send pings from client to server as well as the other direction. If you don't, how do you know when the client disconnected? You're not sending pings you must wait for the underlying connection times out.

In all honesty I'm not sure how other applications that use websockets have architected this, but I opened the issue because my intended use was to have a timer running on the client, and each ping event that came in I would reset the timer on the client. If the timer ran out before a pong was received from the server again, then I know the websocket disconnected unexpectedly.

What I ended up doing is during the websocket connection I have the client send a message that tells the server how long its ping interval is, and I handle the timer on the server. I believe that this websocket library handles the client ping and server pong response automatically, so I don't have to also handle that on the client now and it should recognize the unexpected disconnect if the server doesn't respond to a ping.

Hope that makes sense / clarifies anything for people looking at this later. I'm going to close the issue because I have a workaround for my own use case.

**Comment by ffactory-ofcl** [comment]
I've switched to directly using IOWebSocket and HttpWebSocket because I couldn't find an easy way to set the ping interval. Now the connection's onDone is called immediately after disconnecting and I can execute my reconnect logic.
Works well. 
If you want to see how I did it, look here: https://github.com/ffactory-ofcl/fourinarow-app

I recently added a tcp-like layer of reliablity to acknowledge, queue and reorder messages on client- and server side. The code might not be documented super well because it's my personal project, but feel free to take a look :)
